plugins/woocommerce-admin/client/products/hooks/use-product-attributes.ts

```diff
@@ -0,0 +1,111 @@
+/**
+ * External dependencies
+ */
+import {
+	EXPERIMENTAL_PRODUCT_ATTRIBUTE_TERMS_STORE_NAME,
+	ProductAttribute,
+	ProductAttributeTerm,
+} from '@woocommerce/data';
+import { resolveSelect } from '@wordpress/data';
+import { useCallback, useEffect, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { sift } from '../../utils';
+
+type useProductAttributesProps = {
+	allAttributes: ProductAttribute[];
+	isVariationAttributes?: boolean;
+	onChange: ( attributes: ProductAttribute[] ) => void;
+	productId?: number;
+};
+
+export type EnhancedProductAttribute = ProductAttribute & {
+	terms?: ProductAttributeTerm[];
+	visible?: boolean;
+};
+
+export function useProductAttributes( {
+	allAttributes = [],
+	isVariationAttributes = false,
+	onChange,
+	productId,
+}: useProductAttributesProps ) {
+	const getFilteredAttributes = () => {
+		return isVariationAttributes
+			? allAttributes.filter( ( attribute ) => !! attribute.variation )
+			: allAttributes.filter( ( attribute ) => ! attribute.variation );
+	};
+
+	const [ attributes, setAttributes ] = useState<
+		EnhancedProductAttribute[]
+	>( getFilteredAttributes() );
+	const [ localAttributes, globalAttributes ]: ProductAttribute[][] = sift(
+		attributes,
+		( attr: ProductAttribute ) => attr.id === 0
+	);
+
+	const fetchTerms = useCallback(
+		( attributeId: number ) => {
+			return resolveSelect(
+				EXPERIMENTAL_PRODUCT_ATTRIBUTE_TERMS_STORE_NAME
+			)
+				.getProductAttributeTerms< ProductAttributeTerm[] >( {
+					attribute_id: attributeId,
+					product: productId,
+				} )
+				.then(
+					( attributeTerms ) => {
+						return attributeTerms;
+					},
+					( error ) => {
+						return error;
+					}
+				);
+		},
+		[ productId ]
+	);
+
+	const enhanceAttribute = (
+		globalAttribute: ProductAttribute,
+		terms: ProductAttributeTerm[]
+	) => {
+		return {
+			...globalAttribute,
+			terms: terms.length > 0 ? terms : undefined,
+			options: terms.length === 0 ? globalAttribute.options : [],
+		};
+	};
+
+	const handleChange = ( newAttributes: ProductAttribute[] ) => {
+		const otherAttributes = isVariationAttributes
+			? allAttributes.filter( ( attribute ) => ! attribute.variation )
+			: allAttributes.filter( ( attribute ) => !! attribute.variation );
+		setAttributes( newAttributes );
+		onChange( [ ...otherAttributes, ...newAttributes ] );
+	};
```

Comment:
Could we wrap this in a `useCallback`? I think we should try to memoize any callback functions exposed through a hook as much as possible.